### PR TITLE
Implement CUDA Quantum Kernel from State Vector.

### PR DESCRIPTION
**Pull request by amccaskey**
Enable kernel creation from a state vector. 
```python 
state = np.asarray([.70710678, 0., 0., 0.70710678])
kernel = cudaq.from_state(state)
counts = cudaq.sample(kernel)
```
```cpp
std::vector<std::complex<double>> state{.70710678, 0., 0., 0.70710678};
auto kernel = cudaq::from_state(state);
auto counts = cudaq::sample(*kernel);
```

### CODE REVIEW

File: runtime/cudaq/builder/kernels.cpp [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "kernels.h"
+#include "common/EigenDense.h"
+#include <iostream>
```
Comment (1tnguyen) [inline runtime/cudaq/builder/kernels.cpp:11]: Should we also remove this `#include <iostream>`?

File: runtime/cudaq/builder/kernels.cpp [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+  while (b_and_g > 0) {
+    if (b_and_g & 0b1)
+      sum_of_ones += 1;
+
+    b_and_g = b_and_g >> 1;
+  }
+  return std::pow(-1, sum_of_ones);
+}
+
+std::vector<double> computeAngle(std::vector<double> alpha) {
```
Comment (1tnguyen) [inline runtime/cudaq/builder/kernels.cpp:64]: Does `alpha` scale ~ O(2^N)? Maybe we should use `const std::vector<double>&` to prevent copies.

**Review by 1tnguyen — APPROVED** [review]
Looks great! Just some minor comments :)

**Comment by bmhowe23** [comment]
I am getting `custatevec_fp32_KernelsTester.checkFromState` unit test failures on my workstation from this PR. Is anybody else? It's reporting this:
```
/workspaces/cuda-quantum/unittests/integration/kernels_tester.cpp:141: Failure
The difference between ss[i].real() and state[i].real() is 0.66339194990539552, which exceeds 1e-3, where
ss[i].real() evaluates to 0.95617794990539551,
state[i].real() evaluates to 0.29278599999999999, and
1e-3 evaluates to 0.001.
/workspaces/cuda-quantum/unittests/integration/kernels_tester.cpp:141: Failure
The difference between ss[i].real() and state[i].real() is 0.66339199784088132, which exceeds 1e-3, where
ss[i].real() evaluates to 0.29278600215911865,
state[i].real() evaluates to 0.95617799999999997, and
1e-3 evaluates to 0.001.
```

**Comment by 1tnguyen** [comment]
@bmhowe23 Yea, I can reproduce this as well.

The issue here is that custatevec and qpp backends use different endian conventions when storing the state vector.
In this case, custatevec returned `[0.0, 0.956178, 0.292786, 0.0]` vs. the expected `[0.0, 0.292786, 0.956178, 0.0]`.

I'd also note that this shouldn't affect sampling and observe results since the custatevec backend has accounted for its endianess when computing the expectation value for example.

@amccaskey We probably need a way to make the endian convention of `cudaq::from_state` and `cudaq::get_state` consistent.

**Pull request merged**